fluss-server/src/main/java/org/apache/fluss/server/coordinator/CoordinatorEventProcessor.java

```diff
@@ -1262,20 +1275,58 @@ private void tryProcessCommitLakeTableSnapshot(
+                                                lakeTableSnapshots,
+                                                commitLakeTableSnapshotData
+                                                        .getTableBucketsMaxTieredTimestamp()));
                             } catch (Exception e) {
                                 ApiError error = ApiError.fromThrowable(e);
                                 tableResp.setError(error.error().code(), error.message());
                             }
                         }
+                        callback.complete(response);
+                    } catch (Exception e) {
+                        callback.completeExceptionally(e);
+                    }
+                });
+    }
 
-                        // send notify lakehouse data request to all replicas via coordinator event
-                        coordinatorEventManager.put(
-                                new NotifyLakeTableOffsetEvent(
-                                        lakeTableSnapshots,
-                                        commitLakeTableSnapshotData
-                                                .getTableBucketsMaxTieredTimestamp()));
+    private void handleCommitLakeTableSnapshotV2(
+            CommitLakeTableSnapshotEvent commitLakeTableSnapshotEvent,
+            CompletableFuture<CommitLakeTableSnapshotResponse> callback) {
+        CommitLakeTableSnapshotData commitLakeTableSnapshotData =
+                commitLakeTableSnapshotEvent.getCommitLakeTableSnapshotData();
+        Map<Long, LakeTable.LakeSnapshotMetadata> lakeSnapshotMetadatas =
+                commitLakeTableSnapshotData.getLakeTableSnapshotMetadatas();
+        ioExecutor.execute(
+                () -> {
+                    try {
+                        CommitLakeTableSnapshotResponse response =
+                                new CommitLakeTableSnapshotResponse();
+                        for (Map.Entry<Long, LakeTable.LakeSnapshotMetadata>
+                                lakeSnapshotMetadataEntry : lakeSnapshotMetadatas.entrySet()) {
+                            PbCommitLakeTableSnapshotRespForTable tableResp =
+                                    response.addTableResp();
+                            long tableId = lakeSnapshotMetadataEntry.getKey();
+                            tableResp.setTableId(tableId);
+                            try {
+                                lakeTableHelper.registerLakeTableSnapshotV2(
+                                        tableId, lakeSnapshotMetadataEntry.getValue());
+                                coordinatorEventManager.put(
+                                        new NotifyLakeTableOffsetEvent(
+                                                commitLakeTableSnapshotData.getLakeTableSnapshot(),
+                                                commitLakeTableSnapshotData
+                                                        .getTableBucketsMaxTieredTimestamp()));
```

Comment:
Moving the event-publishing logic into the `forEach` loop would cause the event to be sent multiple times, since the event is intended for all table ids, not per individual table.